### ISSUE TITLE
Prepend imported module file paths with module path in errors

### ISSUE THREAD

**Issue by bufdev** [bufbuild/buf#490]
```
$ buf check breaking --against-input 0.0.0.0:3000/foo/bar/v1:b1-659yVmM4Ut_GZa6Nr-rJW2jT0hmbupCPNFyYr5R28aU=
buf/image/v1/image.proto exists in multiple locations: proto/buf/image/v1/image.proto buf/image/v1/image.proto
```

This actually isn't a bug - this was due to having buf.yaml for bufbuild rely on a pushed module of itself, so the file was in two locations. The error message was just crazy confusing.

This likely gets to my initial intent of external paths: prepend file paths for modules with the module path.

**Comment by TheDrewscriver** [comment]
I would say this is a major issue. This is the problem I am facing:

Even thought the buf.yaml only includes this:
deps:
  - buf.build/beta/googleapis
  - buf.build/grpc-ecosystem/grpc-gateway

When I run 
```
buf mod update -v
DEBUG	get_config	{"duration": "459.9µs"}
DEBUG	grpcclient	dial	{"address": "dns:///api.buf.build"}
DEBUG	buf.alpha.registry.v1alpha1.ResolveService.GetModulePins	{"duration": "624.824732ms", "Client": true, "FailFast": true}
DEBUG	command	{"duration": "730.86471ms"}
```
It adds dependencies for both buf.build/beta and buf.build/googleapis

Contents on buf.lock:
```

# Generated by buf. DO NOT EDIT.
version: v1
deps:
  - remote: buf.build
    owner: beta
    repository: googleapis
    branch: main
    commit: ac0d624de3f746c3b8ef55c74b6f5244
    digest: b1-GubXb4hZ56wTohZYDL-Muute9Jxx8J4TtIYucS30TBM=
    create_time: 2021-08-12T19:53:55.128813Z
  - remote: buf.build
    owner: googleapis
    repository: googleapis
    branch: main
    commit: 6358935a97044c1494fe13d369ac2f3b
    digest: b1-4hxgpEmLuajuxqkZd7NMVnKipt3zqnXaTHsBEKtUOsM=
    create_time: 2021-08-13T15:04:04.653816Z
  - remote: buf.build
    owner: grpc-ecosystem
    repository: grpc-gateway
    branch: main
    commit: d1922677830c448eb65ec590f3abc734
    digest: b1-zmmsl0XlsSF4_BhZe7w9NMdImBCRtMSn68rmn9H1faQ=
    create_time: 2021-08-17T02:45:07.151993Z
```

```
buf lint
threeDS-risk-api.proto:7:8:google/api/annotations.proto exists in multiple locations: google/api/annotations.proto google/api/annotations.proto
```

Shouldn't we only be downloading the referenced dependencies, and not additional dependencies resulting in these kinda errors?

**Comment by amckinney** [comment]
> Shouldn't we only be downloading the referenced dependencies, and not additional dependencies resulting in these kinda errors?

Your issue is slightly different than the one described here. In your case, you're depending on a deprecated module, as described by its module documentation: https://buf.build/beta/googleapis/docs - we need to keep this module around for compatibility reasons, but all users should pin `buf.build/googleapis/googleapis` now.

In order to build your module, you will need the _transitive  closure_ of all of your dependencies. Given that `buf.build/grpc-ecosystem/grpc-gateway` depends on `buf.build/googleapis/googleapis`, both the `beta` and `googleapis` versions are pulled in, which causes this double registration problem.

TL;DR Update your dependency from `buf.build/beta/googleapis` to `buf.build/googleapis/googleapis` and you should be good to go.

**Comment by johanbrandhorst** [comment]
@TheDrewscriver this change was made in the grpc-gateway in https://github.com/grpc-ecosystem/grpc-gateway/commit/d37d6baf3f38053efbdea03a7124afd2c9695fd2. I posted a PSA in the #grpc-gateway channel on Slack, but I wasn't sure how to reach all users. Sorry about the inconvenience.

**Comment by TheDrewscriver** [comment]
@amckinney @johanbrandhorst 
Got it. FIxed, and thanks for the detailed an informative comments. Much appreciated.

**Comment by doriable** [comment]
The first point of discussion here is the idea of prepending module information to the external path. There are a few concerns with this:

- If we simply prepend module information to all representations of external path, then we will have this behaviour in all places that display the external path to users, including `ls-files`. This does not seem like the ideal behaviour for a command like `ls-files`, since users will not be able to depend on them as inputs in automation, etc. Thus, if we are prepending the external path with other metadata, such as module information, we would likely want to either have it as a separate method from `ExternalPath()` so we can use it selectively or have a way to easily parse the prefix.
- Since this particular error is on the storage bucket layer, there is the possibility to extend the storage bucket to take a bucket ID/metadata information, however, this could contain sensitive information (e.g. tmp directory paths, etc.) and is not an ideal way for surfacing clarifying metadata for the user.
- This particular error benefits the most from attaching module information, we may be able to detect this change before the storage layer construction. This is a conversation I had with @amckinney, it would be slightly duplicative logic in that we would be checking the file paths for duplicates at build time, but it would reduce the overhead of changing the other structures and would be the least disruptive to other commands. We could check for duplicate paths during image build: https://github.com/bufbuild/buf/blob/eae4c68ed02f63b4c6c8b4331bf3f543114ba8b0/private/bufpkg/bufimage/bufimagebuild/builder.go#L79-L82

**Comment by amckinney** [comment]
The last approach seems reasonable - the `storage` abstraction can keep its implementation-agnostic feature of detecting duplicate paths, and we can handle this specially for `ImageFiles` with non-nil `ModuleRefs`.

One other approach that builds on this would be to continue leaning on the `storage.NewErrExistsMultipleLocations` error, but export its type so that we can resolve what paths were detected in multiple places. With that, we can map those paths to the `ImageFile` and prepend the `ModuleRef` before it's sent back to the user.

```go
// ErrorExistsMultipleLocations is the error returned if a path exists in multiple locations.
type ErrorExistsMultipleLocations struct {
  Path          string
  ExternalPaths []string
}
```

```go
if v, ok := err.(*ErrorExistsMultipleLocations); ok {
  // Resolve the path from each external path, map it back to the ImageFile, prepend the ModuleRef, etc.
}
```

However, the only benefit that gives us is by reusing the logic already found in the `storage` layer. We would still need to incur all the overhead of mapping each `ExternalPath` back to their `ImageFile` path, so it might not be easier in practice.

I'd suggest going with your last approach, or the suggestion I have above and see which one is simplest.

**Comment by doriable** [comment]
As expressed on the closed PR, we had an offline conversation regarding the performance regression the file check would introduce at the module file set creation time and have made a decision not to pursue this fix. In future reworks of `bufmodule`, we will keep this issue in mind and continue to improve our error messaging to users.
